File: python/ThreeSum.py
class Solution:
    def threeSum(self, nums):
        nums.sort()
        res = []

        for index, value in enumerate(nums):
            if index > 0 and nums[index] == nums[index - 1]:
                continue
            left, right = index + 1, len(nums) - 1
            while left < right:
                currSum = nums[left] + nums[right] + nums[index]
                if currSum > 0:
                    right -= 1
                elif (currSum < 0):
                    left += 1
                else:
                    res.append([value, nums[left], nums[right]])
                    left += 1
                    right -= 1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    while nums[right + 1] == nums[right] and left < right:
                        right -= 1

        return res


arr = [-1, 0, 1, 2, -1, -4]
sol = Solution()
print(sol.threeSum(arr))

File: Javascript/threeSum.js
function threeSum(nums) {
  nums.sort((a, b) => a - b);
  let res = [];
  for (let i = 0; i < nums.length; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) {
      continue;
    }
    let left = i + 1;
    let right = nums.length - 1;

    while (left < right) {
      let currSum = nums[i] + nums[left] + nums[right];
      if (currSum < 0) {
        left++;
      } else if (currSum > 0) {
        right--;
      } else {
        res.push([nums[i], nums[left], nums[right]]);
        left++;
        right--;
        while (left < right && nums[left] === nums[left - 1]) {
          left++;
        }
        while (left < right && nums[right] === nums[right + 1]) {
          right--;
        }
      }
    }
  }
  return res;
}

let arr = [-1, 0, 1, 2, -1, -4];
let val = threeSum(arr);
console.log(val);
